plutus-emulator/src/Wallet/Emulator/Servant.hs

```diff
@@ -0,0 +1,145 @@
+{-# LANGUAGE ConstraintKinds       #-}
+{-# LANGUAGE DeriveAnyClass        #-}
+{-# LANGUAGE DerivingStrategies    #-}
+{-# LANGUAGE FlexibleContexts      #-}
+{-# LANGUAGE FlexibleInstances     #-}
+{-# LANGUAGE GADTs                 #-}
+{-# LANGUAGE LambdaCase            #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
+{-# LANGUAGE OverloadedStrings     #-}
+{-# LANGUAGE TemplateHaskell       #-}
+{-# LANGUAGE TupleSections         #-}
+{-# LANGUAGE TypeApplications      #-}
+module Wallet.Emulator.Servant where
+
+import           Conduit
+import           Control.Concurrent.STM       (atomically)
+import           Control.Concurrent.STM.TVar
+import           Control.Lens                 hiding (index)
+import           Control.Monad.Except
+import           Control.Monad.IO.Class       (MonadIO (..))
+import           Control.Monad.Reader
+import           Control.Monad.State          hiding (state)
+import           Control.Monad.Trans.Resource ()
+import           Data.List                    (partition)
+import           Data.Maybe                   (isNothing)
+import           Data.Traversable             (for)
+import           Servant                      (ServantErr (..))
+
+import           Ledger                       (Block, Blockchain, Slot (..), Tx (..), unspentOutputs)
+import qualified Ledger.Index                 as Index
+import qualified Ledger.Interval              as Interval
+
+-- | A pool of transactions which have yet to be validated.
+type TxPool = [Tx]
+
+data ServerState = ServerState {
+    _chainNewestFirst :: Blockchain, -- ^ The current chain, with the newest transactions first in the list.
+    _txPool           :: TxPool, -- ^ The pool of pending transactions.
+    _index            :: Index.UtxoIndex -- ^ The UTxO index, used for validation.
+} deriving (Show)
+
+makeLenses ''ServerState
+
+type MonadServer m = (MonadReader (TVar ServerState) m, MonadError ServantErr m, MonadIO m)
```

Comment:
Couldn't we define a class that abstracts over what `MonadReader (TVar ServerState) m` provides? i.e. it lets you read or modify things, and it's up to the implementation to do the tvar stuff. Then I think we could provide a pure version that's just implemented with `State ServerState`.

In fact, I think what I'm proposing is something like this:
```
newtype TVarReader r a = TVarReader { unTVarReader :: ReaderT r IO a }

instance MonadState r (TVarReader r) where
    state f = TVarReader $ do
        r <- ask
        atomically $ do
            s <- readTVar r
            let (a, s') = f s
            writeTVar r s'
            pure a
```
i.e. we can just code against `MonadState ServerState` and then write a monad that backs that with a `TVar`. Does that help with the IO problem?

Reply:
This is basically the "global state" monad, and obviously you need to be a little careful with it since e.g. you can't do local state changes and revert them safely etc.

Reply:
Regardless of this, we probably want to have a class that exposes the *operations* of the chain client rather than just raw access to its state, so we can swap our implementations.